exporter/azuredataexplorerexporter/metricsdata_to_adx.go

```diff
@@ -0,0 +1,332 @@
+package azuredataexplorerexporter
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"os"
+	"strconv"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+	"go.uber.org/zap"
+)
+
+/*
+A converter package that converts and marshals data to be written to ADX metrics tables
+*/
+const (
+	hostkey = "host.key"
+	// Indicates the sum that is used in both summary and in histogram
+	sumsuffix = "sum"
+	// Count used in summary , histogram and also in exponential histogram
+	countsuffix = "count"
+	// Indicates the sum that is used in both summary and in histogram
+	sumdescription = "(Sum total of samples)"
+	// Count used in summary , histogram and also in exponential histogram
+	countdescription = "(Count of samples)"
+)
+
+// This is derived from the specification https://opentelemetry.io/docs/reference/specification/metrics/datamodel/
+type AdxMetric struct {
+	Timestamp string //The timestamp of the occurance. A metric is measured at a point of time. Formatted into string as RFC3339
+	//Including name, the Metric object is defined by the following properties:
+	MetricName        string                 //Name of the metric field
+	MetricType        string                 // The data point type (e.g. Sum, Gauge, Histogram ExponentialHistogram, Summary)
+	MetricUnit        string                 // The metric stream’s unit
+	MetricDescription string                 //The metric stream’s description
+	MetricValue       float64                // the value of the metric
+	MetricAttributes  map[string]interface{} // JSON attributes that can then be parsed. Extrinsic properties
+	//Additional properties
+	Host               string                 // the hostname for analysis of the metric. Extracted from https://opentelemetry.io/docs/reference/specification/resource/semantic_conventions/host/
+	ResourceAttributes map[string]interface{} // The originating Resource attributes. Refer https://opentelemetry.io/docs/reference/specification/resource/sdk/
+}
+
+/*
+	Convert the pMetric to the type ADXMetric , this matches the scheme in the OTELMetric table in the database
+*/
+
+func mapToAdxMetric(res pcommon.Resource, md pmetric.Metric, scopeattrs map[string]interface{}, logger *zap.Logger) []*AdxMetric {
+	logger.Debug("Entering processing of toAdxMetric function")
+	// default to collectors host name. Ignore the error here. This should not cause the failure of the process
+	host, _ := os.Hostname()
+	commonFields := map[string]interface{}{}
+	resourceattrs := res.Attributes().AsRaw()
+	res.Attributes().Range(func(k string, v pcommon.Value) bool {
+		switch k {
+		case hostkey:
+			host = v.StringVal()
+		default:
+			commonFields[k] = v.AsString()
+		}
+		return true
+	})
+	switch md.DataType() {
+	case pmetric.MetricDataTypeGauge:
+		dataPoints := md.Gauge().DataPoints()
+		adxMetrics := make([]*AdxMetric, dataPoints.Len())
+		for gi := 0; gi < dataPoints.Len(); gi++ {
+			dataPoint := dataPoints.At(gi)
```

Comment:
Another chunk of very similar code for each data type. Please merge and minimize as much as possible.

Reply:
Hello @AsafMah , this is a good point. Unfortunately , though the methods are the same the internal objects where we access these methods are different , for example we have HistogramDataPointSlice , GaugeDataPointSlice etc. though these are all part of metrics. There seems to be no common hierarchy to make these common. We were checking other exporters too , they seem to have similar repetition

https://github.com/ag-ramachandran/opentelemetry-collector-contrib/blob/9d5b9303932630580037f40d0d72f0b613940269/exporter/splunkhecexporter/metricdata_to_splunk.go#L91